plugins/com.google.cloud.tools.eclipse.appengine.facets/src/com/google/cloud/tools/eclipse/appengine/facets/FacetUtil.java

```diff
@@ -0,0 +1,211 @@
+import org.eclipse.wst.common.frameworks.datamodel.DataModelFactory;
+import org.eclipse.wst.common.frameworks.datamodel.IDataModel;
+import org.eclipse.wst.common.project.facet.core.IFacetedProject;
+import org.eclipse.wst.common.project.facet.core.IProjectFacetVersion;
+
+/**
+ * Utility class for processing facets.
+ */
+@SuppressWarnings("restriction") // For IDependencyGraph
+public class FacetUtil {
+  public static final Logger logger = Logger.getLogger(FacetUtil.class.getName());
+  
+
+  /**
+   * Adds an install action for the {@code javaFacet} to the {@code facetInstallSet} if the
+   * {@code javaFacet} does not already exist in {@code facetedProject}.
+   *
+   * @return true is the facet action was added to {@code facetInstallSet} and false if {@code javaFacet}
+   *   already exists in {@code facetedProject}
+   * @throws CoreException if {@code javaFacet} is not a Java facet
+   */
+ public static boolean addJavaFacetToBatch(IProjectFacetVersion javaFacet, IFacetedProject facetedProject,
+     Set<IFacetedProject.Action> facetInstallSet) throws CoreException {
+   if (!JavaFacet.FACET.getId().equals(javaFacet.getProjectFacet().getId())) {
+     throw new CoreException(StatusUtil.error(FacetUtil.class, javaFacet.toString() + " is not a Java facet"));
+   }
+   
+   if (facetedProject.hasProjectFacet(javaFacet)) {
+     return false;
+   }
+
+   JavaFacetInstallConfig javaConfig = new JavaFacetInstallConfig();
+   List<IPath> sourcePaths = new ArrayList<>();
+
+   IProject project = facetedProject.getProject();
+   // TODO: https://github.com/GoogleCloudPlatform/google-cloud-eclipse/issues/490
+   if (project.getFolder("src/main/java").exists()) {
+     sourcePaths.add(new Path("src/main/java"));
+   }
+
+   if (project.getFolder("src/test/java").exists()) {
+     sourcePaths.add(new Path("src/test/java"));
+   }
+
+   javaConfig.setSourceFolders(sourcePaths);
+   facetInstallSet.add(new IFacetedProject.Action(
+       IFacetedProject.Action.Type.INSTALL, javaFacet, javaConfig));
+   return true;
+ }
+
+ /**
+  * Adds an install action for the {@code webFacet} to the {@code facetInstallSet} if the
+  * {@code webFacet} does not already exist in {@code facetedProject}.
+  * 
+  * @return true is the facet action was added to {@code facetInstallSet} and false if {@code webFacet}
+  *   already exists in {@code facetedProject}
+  * @throws CoreException if {@code webFacet} is not a Web facet
+  */
+ public static boolean addWebFacetToBatch(IProjectFacetVersion webFacet, IFacetedProject facetedProject,
+     Set<IFacetedProject.Action> facetInstallSet) throws CoreException {
+   if (!WebFacetUtils.WEB_FACET.getId().equals(webFacet.getProjectFacet().getId())) {
+     throw new CoreException(StatusUtil.error(FacetUtil.class, webFacet.toString() + " is not a Web facet"));
+   }
+
+   if (facetedProject.hasProjectFacet(webFacet)) {
+     return false;
+   }
+
+   String webAppDirectory = "src/main/webapp";
```

Comment:
It would be good to cause`WebProjectUtil.getWebInfFolder()` to share this functionality.